<launch>
  <arg name="ns" default=""/>
  <arg name="use_gazebo" default="false"/>
  <arg name="marker_id" default="1"/>
  <arg name="robot_localization" default="0"/>
  <arg name="hector_mapping" default="0"/>
  <arg name="octomap" default="0"/>
  <arg name="gmapping" default="0"/>

  <!-- ################################################# -->
  <!-- TRANSFORMS -->
  <!-- ################################################# -->
  
  <group if="$(arg octomap)"> <!-- only octomap cannot create its own map frame, thus we pin it to odom -->
    <node pkg="tf" type="static_transform_publisher" name="map_odom" args="0 0 0 0 0 0 1 map $(arg ns)/odom 100"/>
  </group>
  
  <!-- ################################################# -->
  <!-- ROBOT LOCALIZATION (for octomapping (aka "mapping with known poses")) -->
  <!-- ################################################# -->

  <group if="$(arg robot_localization)">
    <node pkg="robot_localization" type="ekf_localization_node" name="odom_twb_ekf" clear_params="true">
      <param name="frequency" value="50"/>
      <param name="map_frame" value="$(arg ns)/map"/>
      <param name="odom_frame" value="$(arg ns)/odom"/>
      <param name="base_link_frame" value="$(arg ns)/base_link"/>

      <param name="two_d_mode" value="true"/>
      <param name="publish_tf" value="true"/>
      <param name="print_diagnostics" value="true"/>

      <!-- values is x, y, z, roll, pitch, yaw, vx, vy, vz, vroll, vpitch, vyaw, ax, ay, az. -->
      <param name="odom0" value="odom"/>
      <rosparam param="odom0_config">[true, true, false, false, false, true, false, false, false, false, false, false, false, false, false]</rosparam>
      <param name="odom0_differential" value="true"/>
      <param name="odom0_relative" value="true"/>

      <param name="odom1" value="/tracking/$(arg marker_id)"/>
      <rosparam param="odom1_config">[true, true, false, false, false, true, false, false, false, false, false, false, false, false, false]</rosparam>
      <param name="odom1_differential" value="false"/>
      <param name="odom1_relative" value="false"/>
    </node>
  </group>

  <!-- ################################################# -->
  <!-- HECTOR MAPPING -->
  <!-- ################################################# -->
  <group if="$(arg hector_mapping)">
    <node pkg="hector_mapping" type="hector_mapping" name="hector_mapping" output="screen">
      <param name="scan_topic" value="laser"/>
      <param name="tf_map_scanmatch_transform_frame_name" value="/$(arg ns)/tf_map_scanmatch_transform_frame_name" />

      <param name="map_frame" value="map"/>
      <param name="odom_frame" value="$(arg ns)/odom"/>
      <param name="base_frame" value="$(arg ns)/base_link"/>

      <param name="map_resolution" value="0.025"/>
      <param name="map_size" value="5000"/>

      <param name="pub_map_odom_transform" value="true"/>
      <param name="map_pub_period" value="0.1"/>
    </node>
  </group>

  <!-- ################################################# -->
  <!-- OCTOMAP (aka mapping with known poses)-->
  <!-- ################################################# -->

  <group if="$(arg octomap)">
    <node pkg="amiro_slam" type="ros_laser_scan_to_ros_pointcloud2" output="screen" name="laser_to_pointcloud2">
      <param name="laserscan_topic" value="/$(arg ns)/laser"/>
      <param name="pointcloud2_topic" value="/$(arg ns)/pointcloud2"/>
    </node>
    <node pkg="octomap_server" type="octomap_server_node" name="octomap_server">
      <param name="resolution" value="0.0500"/>
      <param name="frame_id" value="map"/>
      <param name="base_frame_id" value="$(arg ns)/base_link"/>
      <param name="sensor_model/max_range" value="3"/> <!-- a bit less than the full range, otherwise the floor will be mapped -->
      <remap from="cloud_in" to="pointcloud2"/>
      <remap from="projected_map" to="map"/>
    </node>
  </group>

  <!-- ################################################# -->
  <!-- GMAPPING -->
  <!-- ################################################# -->

  <group if="$(arg gmapping)">
    <node pkg="gmapping" type="slam_gmapping" name="gmapping_thing" output="screen">
      <remap from="scan" to="laser"/>
      <param name="base_frame" value="$(arg ns)/base_link"/>
      <param name="odom_frame" value="$(arg ns)/odom"/>
      <param name="map_update_interval" value="0.5"/>
      <!-- <param name="minimumScore" value="50"/> -->
      <param name="delta" value="0.025"/>
      <param name="minimumScore" value="0"/>
      <param name="particles" value="50"/>
      <param name="kernelSize" value="1"/>
    </node>
  </group>

</launch>
